Mathlib/Combinatorics/SimpleGraph/Metric.lean

```diff
@@ -43,105 +41,179 @@ variable {V : Type*} (G : SimpleGraph V)
 
 /-! ## Metric -/
 
+section edist
+
+/--
+The extended distance between two vertices is the length of the shortest walk between them.
+It is `⊤` if no such walk exists.
+-/
+noncomputable def edist (u v : V) : ℕ∞ :=
+  ⨅ w : G.Walk u v, w.length
+
+variable {G} {u v w : V}
+
+theorem edist_eq_sInf : G.edist u v = sInf (Set.range fun w : G.Walk u v ↦ (w.length : ℕ∞)) := rfl
+
+protected theorem Reachable.exists_walk_length_eq_edist (hr : G.Reachable u v) :
+    ∃ p : G.Walk u v, p.length = G.edist u v :=
+  csInf_mem <| Set.range_nonempty_iff_nonempty.mpr hr
 
-/-- The distance between two vertices is the length of the shortest walk between them.
-If no such walk exists, this uses the junk value of `0`. -/
+protected theorem Connected.exists_walk_length_eq_edist  (hconn : G.Connected) (u v : V) :
+    ∃ p : G.Walk u v, p.length = G.edist u v :=
+  (hconn u v).exists_walk_length_eq_edist
+
+theorem edist_le (p : G.Walk u v) :
+    G.edist u v ≤ p.length :=
+  sInf_le ⟨p, rfl⟩
+protected alias Walk.edist_le := edist_le
+
+@[simp]
+theorem edist_eq_zero_iff :
+    G.edist u v = 0 ↔ u = v := by
+  apply Iff.intro <;> simp [edist, ENat.iInf_eq_zero]
+
+theorem edist_self : edist G v v = 0 :=
+  edist_eq_zero_iff.mpr rfl
+
+theorem edist_pos_of_ne (hne : u ≠ v) :
+    0 < G.edist u v :=
+  pos_iff_ne_zero.mpr <| edist_eq_zero_iff.ne.mpr hne
+protected alias Ne.edist_pos := edist_pos_of_ne
+
+lemma edist_eq_top_of_not_reachable (h : ¬G.Reachable u v) :
+    G.edist u v = ⊤ := by
+  simp [edist, not_reachable_iff_isEmpty_walk.mp h]
+
+theorem reachable_of_edist_ne_top (h : G.edist u v ≠ ⊤) :
+    G.Reachable u v :=
+  not_not.mp <| edist_eq_top_of_not_reachable.mt h
+
+protected theorem Connected.edist_triangle (hconn : G.Connected) :
+    G.edist u w ≤ G.edist u v + G.edist v w := by
+  obtain ⟨p, hp⟩ := hconn.exists_walk_length_eq_edist u v
+  obtain ⟨q, hq⟩ := hconn.exists_walk_length_eq_edist v w
+  rw [← hp, ← hq, ← Nat.cast_add, ← Walk.length_append]
+  apply edist_le
+
+theorem edist_comm : G.edist u v = G.edist v u := by
+  rw [edist_eq_sInf, ← Set.image_univ, ← Set.image_univ_of_surjective Walk.reverse_surjective,
+    ← Set.image_comp, Set.image_univ, Function.comp_def]
+  simp_rw [Walk.length_reverse, ← edist_eq_sInf]
+
+lemma exists_walk_of_edist_ne_top (h : G.edist u v ≠ ⊤) :
+    ∃ p : G.Walk u v, p.length = G.edist u v :=
+  (reachable_of_edist_ne_top h).exists_walk_length_eq_edist
```

Comment:
```
lemma exists_walk_of_edist_eq_coe {k : ℕ} (h : G.edist u v = k) :
    ∃ p : G.Walk u v, p.length = k :=
  have : G.edist u v ≠ ⊤ := by rw [h]; exact ENat.coe_ne_top _
  have ⟨p, hp⟩ := exists_walk_of_edist_ne_top this
  ⟨p, Nat.cast_injective (hp.trans h)⟩
```
this seems helpful, I think.

Also a statement that edist isn't top iff the vertices are reachable (as an iff) would be good